src/tasks/block.rs

```diff
@@ -1,329 +1,352 @@
+    ///
+    /// # Returns
+    ///
+    /// An `Instant` representing the deadline.
+    pub fn calculate_deadline(&self) -> Instant {
+        let now = SystemTime::now();
+        let unix_seconds = now.duration_since(UNIX_EPOCH).expect("Time went backwards").as_secs();
 
-    // add a buffer to the beginning of the block slot.
-    fn secs_to_next_target(&self) -> u64 {
-        self.secs_to_next_slot() + self.config.target_slot_time
+        Instant::now().checked_add(Duration::from_secs(unix_seconds)).unwrap()
     }
 
-    /// Spawn the block builder task, returning the inbound channel to it, and
-    /// a handle to the running task.
-    pub fn spawn(mut self, outbound: mpsc::UnboundedSender<InProgressBlock>) -> JoinHandle<()> {
-        tokio::spawn(
-            async move {
-                loop {
-                    // sleep the buffer time
-                    tokio::time::sleep(Duration::from_secs(self.secs_to_next_target())).await;
-                    info!("beginning block build cycle");
-
-                    // Build a block
-                    let mut in_progress = InProgressBlock::default();
-                    self.get_transactions(&mut in_progress).await;
-                    self.get_bundles(&mut in_progress).await;
-
-                    // Filter confirmed transactions from the block
-                    self.filter_transactions(&mut in_progress).await;
-
-                    // submit the block if it has transactions
-                    if !in_progress.is_empty() {
-                        debug!(txns = in_progress.len(), "sending block to submit task");
-                        let in_progress_block = std::mem::take(&mut in_progress);
-                        if outbound.send(in_progress_block).is_err() {
-                            error!("downstream task gone");
-                            break;
-                        }
-                    } else {
-                        debug!("no transactions, skipping block submission");
-                    }
-                }
+    /// Creates an `AlloyDB` instance from the rollup provider.
+    ///
+    /// # Returns
+    ///
+    /// An `Option` containing the wrapped database or `None` if an error occurs.
+    async fn create_db(&self) -> Option<AlloyDatabaseProvider> {
+        let latest = match self.ru_provider.get_block_number().await {
+            Ok(block_number) => block_number,
+            Err(e) => {
+                tracing::error!(error = %e, "failed to get latest block number");
+                return None;
             }
-            .in_current_span(),
-        )
+        };
+        let alloy_db: AlloyDB<Ethereum, RuProvider> =
+            AlloyDB::new(self.ru_provider.clone(), BlockId::from(latest));
+        let wrapped_db: WrapDatabaseAsync<AlloyDB<Ethereum, RuProvider>> =
+            WrapDatabaseAsync::new(alloy_db).unwrap_or_else(|| {
+                panic!("failed to acquire async alloy_db; check which runtime you're using")
```

Comment:
this errorr only occurs on current thread runtimes, so we can actually rule it out entirely by setting the runtime flavor in the main macro [link](https://docs.rs/tokio/latest/tokio/attr.main.html)